lib/runtime/src/ops/linear.cc

```diff
@@ -25,1063 +29,258 @@ using Legion::TaskLauncher;
-  for (int i = 1; i < pc.nDims - 1; i++) {
-    pc.dim[i] = 1;
-  }
-  int start_idx = std::rand() % (total_devices - num_par_c * num_par_b + 1);
-  start_idx = start_idx - start_idx % num_par_c;
-  for (int i = 0; i < num_par_c * num_par_b; i++) {
-    pc.device_ids[i] = start_idx + i;
-  }
-  return pc;
-}
 
-bool Linear::get_int_parameter(PMParameter para, int *value) const {
-  switch (para) {
-    case PM_ACTI:
-      *value = (int)activation;
-      return true;
-    default:
-      return Op::get_int_parameter(para, value);
-  }
+  int in_dim = input.shape.at(ff_dim_t{0}) + 1;
+  int out_dim = output.shape.at(ff_dim_t{0}) + 1;
+  int batch_size = output.shape.get_volume() / out_dim;
+
+  return profile(backward_kernel,
+                 profiling,
+                 "[Linear] backward_time = %.2lfms\n",
+                 per_device_state,
+                 input.get_float_ptr(),
+                 input_grad.get_float_ptr(),
+                 output.get_float_ptr(),
+                 output_grad.get_float_ptr(),
+                 weight.get_float_ptr(),
+                 weight_grad.get_float_ptr(),
+                 bias_ptr,
+                 in_dim,
+                 out_dim,
+                 batch_size);
 }
 
-bool Linear::is_valid_parallel_config(FFModel const &ff,
-                                      ParallelConfig const &pc) const {
-  if (!ff.config.enable_parameter_parallel) {
-    return Op::is_valid_parallel_config(ff, pc);
-  }
-  // Support data and parameter parallel
-  if (pc.nDims != outputs[0]->num_dims) {
-    return false;
-  }
-  for (int i = 1; i < pc.nDims - 1; i++) {
-    if (pc.dim[i] != 1) {
-      return false;
-    }
-  }
-  return true;
+static void backward_task(Task const *task,
+                          std::vector<PhysicalRegion> const &regions,
+                          Context ctx,
+                          Runtime *runtime) {
+  TaskArgumentAccessor acc(task, regions, ctx, runtime);
+  backward_task_impl(acc);
 }
 
-bool Linear::measure_operator_cost(Simulator *sim,
-                                   MachineView const &mv,
-                                   CostMetrics &cost_metrics) const {
-  ParallelTensorBase sub_output, sub_input;
-  if (!outputs[0]->get_sub_tensor(mv, sub_output)) {
-    return false;
-  }
-  if (!inputs[0]->get_sub_tensor(mv, sub_input)) {
-    return false;
-  }
-  int input_c = sub_input.dims[0].size;
-  int input_n = sub_input.get_volume() / input_c;
-  int output_c = sub_output.dims[0].size;
-  int output_n = sub_output.get_volume() / output_c;
-  LinearMeta *m = sim->linear_meta;
-  m->activation = activation;
-  m->input_type = inputs[0]->data_type;
-  m->weight_type = this->data_type;
-  m->output_type = outputs[0]->data_type;
-  assert(m->profiling == false);
+CostMetrics measure_operator_cost(SimEnvFactory const &sim_factory,
+                                  LinearAttrs const &attrs,
+                                  ParallelTensorShape const &input_shape,
```

Comment:
I think these should always be `InputParallelTensorDesc` instead of `ParallelTensorShape`